### PR TITLE
further typing in the simplifier

### PR DESCRIPTION
**Pull request by kroening**
This improves type safety.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0829d56f4d7d3338244fce74838d6404f85f3ab6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122998356

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1e039665fb076c32f33e1d76be2251365299eb99).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123203136
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 72c3c0448d32edcb0bead3d53dc3396a0cb12e95).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123206084
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 342fd4388245cf4832361e4452d6d84aa2630bee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123207237

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a8c979a11dc7dfd72e463fd6d4dc8858425f86ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123237648

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -656,28 +656,33 @@ simplify_exprt::simplify_typecast(const typecast_exprt &expr)
...
-     (expr.op().op0().is_constant() &&
-      to_constant_expr(expr.op().op0()).get_value() == ID_NULL)))
-  {
-    auto sub_size = pointer_offset_size(to_pointer_type(op_type).subtype(), ns);
-    if(sub_size.has_value())
+    expr.op().operands().size() == 2)
+  {
+    const auto &op_plus_expr = to_plus_expr(expr.op());
+
+    if(((op_plus_expr.op0().id() == ID_typecast &&
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp:663]: There is still an `op0()` call here; not sure if that is intentional.

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -1139,16 +1145,19 @@ simplify_exprt::simplify_dereference(const dereference_exprt &expr)
     pointer.id() == ID_plus && pointer.operands().size() == 2 &&
     to_plus_expr(pointer).op0().id() == ID_address_of)
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp:1146]: Another one.

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -1139,16 +1145,19 @@ simplify_exprt::simplify_dereference(const dereference_exprt &expr)
...
   {
-    const address_of_exprt &address_of=
-      to_address_of_expr(pointer.op0());
+    const auto &pointer_plus_expr = to_plus_expr(pointer);
+
+    const address_of_exprt &address_of =
+      to_address_of_expr(pointer_plus_expr.op0());
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp:1151]: Another.

File: src/util/simplify_expr.cpp [hunk 4]
```diff
@@ -1139,16 +1145,19 @@ simplify_exprt::simplify_dereference(const dereference_exprt &expr)
...
+
     if(address_of.object().id()==ID_index)
     {
       const index_exprt &old=to_index_expr(address_of.object());
       if(old.array().type().id() == ID_array)
       {
         index_exprt idx(
           old.array(),
-          pointer_offset_sum(old.index(), pointer.op1()),
+          pointer_offset_sum(old.index(), pointer_plus_expr.op1()),
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp:1160]: Another

File: src/util/simplify_expr.cpp [hunk 5]
```diff
@@ -1367,31 +1373,31 @@ simplify_exprt::resultt<> simplify_exprt::simplify_object(const exprt &expr)
...
           else if(
             typecast_op.id() == ID_plus && typecast_op.operands().size() == 2 &&
-            typecast_op.op0().id() == ID_typecast &&
-            to_typecast_expr(typecast_op.op0()).op().id() == ID_address_of)
+            to_plus_expr(typecast_op).op0().id() == ID_typecast &&
+            to_typecast_expr(to_plus_expr(typecast_op).op0()).op().id() ==
+              ID_address_of)
           {
-            return cand.op0().op0().op0();
+            return to_typecast_expr(to_plus_expr(typecast_op).op0()).op();
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp:1380]: Another.

File: src/util/simplify_expr.cpp [hunk 6]
```diff
@@ -1885,7 +1892,7 @@ simplify_exprt::simplify_byte_update(const byte_update_exprt &expr)
     expr.value().type() == to_byte_update_expr(expr.op()).value().type())
   {
     auto tmp = expr;
-    tmp.set_op(expr.op().op0());
+    tmp.set_op(to_byte_update_expr(expr.op()).op0());
```
Comment (martin-cs) [inline src/util/simplify_expr.cpp]: Another `op0()`

**Review by martin-cs — COMMENTED** [review]
I'm not sure if this is part of the "removing `opX()`" patch set.  If it is, then it needs work, if not, then it seems fine.

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (kroening) [reply src/util/simplify_expr.cpp:663]: Yes; note that these are run-time checked in a `multi_ary_exprt`, so this is memory safe.

File: src/util/simplify_expr.cpp (hunk 6 above)
Comment (kroening) [reply src/util/simplify_expr.cpp]: fixed!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d85bafd790c97b625873ae691e290d067007de4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124099078

File: src/util/simplify_expr_pointer.cpp [hunk 7]
```diff
@@ -421,48 +422,69 @@ simplify_exprt::simplify_pointer_offset(const unary_exprt &expr)
...
+
+  // skip over the typecast
   if(tmp0.id()==ID_typecast)
-    tmp0=expr.op0().op0();
-  if(tmp0.op0().id()==ID_index &&
-     to_index_expr(tmp0.op0()).index().is_zero())
-    tmp0=address_of_exprt(to_index_expr(tmp0.op0()).array());
+    tmp0 = to_typecast_expr(tmp0).op();
+
+  PRECONDITION(tmp0.id() == ID_address_of);
```
Comment (peterschrammel) [inline src/util/simplify_expr_pointer.cpp:437]: That's also checked in `to_address_of_expr`

File: src/util/simplify_expr_pointer.cpp [hunk 8]
```diff
@@ -421,48 +422,69 @@ simplify_exprt::simplify_pointer_offset(const unary_exprt &expr)
...
-  INVARIANT(tmp0.id() == ID_address_of, "id must be ID_address_of");
-  INVARIANT(tmp1.id() == ID_address_of, "id must be ID_address_of");
-
-  if(tmp0.operands().size()!=1)
-    return unchanged(expr);
-  if(tmp1.operands().size()!=1)
-    return unchanged(expr);
+    tmp1 = to_typecast_expr(tmp1).op();
+
+  PRECONDITION(tmp1.id() == ID_address_of);
```
Comment (peterschrammel) [inline src/util/simplify_expr_pointer.cpp:455]: same here

**Review by peterschrammel — APPROVED** [review]

File: src/util/simplify_expr_pointer.cpp (hunk 7 above)
Comment (kroening) [reply src/util/simplify_expr_pointer.cpp:437]: yes, left deliberately, to document the contract of this function.

**Pull request merged**